### PR TITLE
Add support for repeatable annotations

### LINKED ISSUE #5092: Support @Repeatable Java annotations
See https://github.com/scala/scala/pull/6846

### PR DESCRIPTION
**Pull request by Kordyjan**
Fixes #5092

Before this PR, using multiple annotations of the same type on any symbol would generate incorrect bytecode; this can lead to runtime exceptions when reflection is used.
This PR makes the compiler check if they are multiple annotations of the same type and raise a compile error or aggregate them if their type supports it.

### CODE REVIEW

**Comment by hepin1989** [comment]
refs: https://github.com/scala/scala/pull/6846

**Comment by smarter** [comment]
@sjrd, will you be able to review this?

**Comment by sjrd** [comment]
Hum, probably not. I didn't notice I had been assigned, and I'm not sure why either: it's not at all my area of expertise.

**Comment by smarter** [comment]
@liufengyun would you be available to review this?

**Comment by liufengyun** [comment]
> @liufengyun would you be available to review this?

Yes, I can have a look.

File: compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    if (!annotations.isEmpty) then
+      tree.symbol.annotations = aggregateAnnotations(tree.symbol.annotations)
+    tree
+
+  private def aggregateAnnotations(annotations: Seq[Annotation])(using Context): List[Annotation] =
+    val annsByType = annotations.groupBy(_.symbol)
+    annsByType.flatMap {
+      case (_, a :: Nil) => a :: Nil
+      case (sym, anns) if sym.derivesFrom(defn.ClassfileAnnotationClass) =>
+        sym.annotations.find(_ matches defn.JavaRepeatableAnnot).flatMap(_.argumentConstant(0)) match
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala:32]:
```suggestion
        sym.annotations.getAnnotation(defn.JavaRepeatableAnnot).flatMap(_.argumentConstant(0)) match
```

File: compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          case Some(Constant(containerTpe: Type)) =>
+            val clashingAnns = annsByType.getOrElse(containerTpe.classSymbol, Nil)
+            if !clashingAnns.isEmpty then
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala:35]:
```suggestion
            if clashingAnns.nonEmpty then
```

File: compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+              // this is the same error javac would raise in this case
+              val pos = clashingAnns.map(_.tree.srcPos).minBy(_.line)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala:37]: I think reporting the first is already good enough:

```suggestion
              val pos = clashingAnns.head.tree.srcPos
```

File: compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+              report.error("Container must not be present at the same time as the element it contains", pos)
+              Nil
+            else
+              val aggregated = JavaSeqLiteral(anns.map(_.tree).toList, TypeTree(sym.typeRef))
+              Annotation(containerTpe, NamedArg("value".toTermName, aggregated)) :: Nil
+          case _ =>
+            val pos = anns.map(_.tree.srcPos).sortBy(_.line).apply(1)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala:44]:
```suggestion
            val pos = anns.head.tree.srcPos
```

File: compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  override def phaseName = "repeatableAnnotations"
+
+  override def transformTypeDef(tree: TypeDef)(using Context): Tree = transformDef(tree)
+  override def transformValDef(tree: ValDef)(using Context): Tree = transformDef(tree)
+  override def transformDefDef(tree: DefDef)(using Context): Tree = transformDef(tree)
+
+  private def transformDef(tree: DefTree)(using Context) =
+    val annotations = tree.symbol.annotations
+    if (!annotations.isEmpty) then
+      tree.symbol.annotations = aggregateAnnotations(tree.symbol.annotations)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala:24]: Usually, we use `SymTransformer` to change symbols. I think here it's fine to do so as the change is not supposed to be visible in a separately compiled module.

File: compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala (hunk 1 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/RepeatableAnnotations.scala:32]: Minor: I'd pattern match on `sym.annotations.getAnnotation(defn.JavaRepeatableAnnot)`, and then call `argumentConstant(0)` in the `Some` branch. This is more clear.

**Review by liufengyun — APPROVED** [review]
Otherwise, LGTM 👍

**Pull request merged**